[cmdpal] Support invoke command in Bookmarks extension

moooyo opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request
![image](https://github.com/user-attachments/assets/e6947bd7-6532-4a3f-bdbc-35a180654c45)
![image](https://github.com/user-attachments/assets/0889ac27-5d6a-49be-8b76-3c4dba1eada6)

* Support invoke command
* Add BookmarkType enum to categorize bookmarks
* Use JsonSerializer to serialize all i18n string to avoid invalid character.
* Add IconHelper to create icon for bookmark by type.
* Forward compatibility. No issue if we upgrade from old bookmark ext.


<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist

- [x] **Closes:** #38700
- [x] **Communication:** I've discussed this with core contributors already. If work hasn't been agreed, this work might be rejected
- [x] **Tests:** Added/updated and all pass
- [x] **Localization:** All end user facing strings can be localized
- [ ] **Dev docs:** Added/updated
- [ ] **New binaries:** Added on the required places
   - [ ] [JSON for signing](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ESRPSigning_core.json) for new binaries
   - [ ] [WXS for installer](https://github.com/microsoft/PowerToys/blob/main/installer/PowerToysSetup/Product.wxs) for new binaries and localization folder
   - [ ] [YML for CI pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ci/templates/build-powertoys-steps.yml) for new test projects
   - [ ] [YML for signed pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/release.yml)
- [ ] **Documentation updated:** If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/windows-uwp/tree/docs/hub/powertoys) and link it here: #xxx

<!-- Provide a more detailed description of the PR, other things fixed or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments

<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed

Commit:
init

Commit:
Change structure

Commit:
Fix icon issues

Commit:
Fix cmd issue

Commit:
Fix i18n issue

Commit:
Remove unused field.

Commit:
Folder not web need directoryPage and OpenInShellCommand

Commit:
Add subtitle for shell command

Commit:
Fix typo issue

moooyo commented:
/azp run

htcfreek commented on src/modules/cmdpal/ext/Microsoft.CmdPal.Ext.Bookmark/Command/ShellCommand.cs:
```diff
@@ -0,0 +1,81 @@
+
+        Name = name;
+
+        // Icon = new IconInfo(IconFromUrl(Value, type));
```
Left over? Otherwise please add description to comment.

htcfreek commented on src/modules/cmdpal/ext/Microsoft.CmdPal.Ext.Bookmark/Bookmarks.cs:
```diff
@@ -41,4 +42,11 @@ public static void WriteToFile(string path, Bookmarks data)
         File.WriteAllText(BookmarksCommandProvider.StateJsonPath(), jsonString);
     }
+
+    public enum BookmarkType
```
Why have two different type enums? This one vs the one in `Models`. Makes no sense and is a mistake source.

moooyo replied:
Oh... forgot to remove it.

htcfreek commented on src/modules/cmdpal/ext/Microsoft.CmdPal.Ext.Bookmark/Helpers/IconHelper.cs:
```diff
@@ -0,0 +1,77 @@
+            }
+            catch (UriFormatException)
+            {
+                // return "🔗";
```
Why is this commented? We could return an arlert sign instead.

Better would be a comment like "Catch without action" or add logging.

moooyo replied:
Added a logging here.

Use default icon now, if we have a new icon for error path, we can use it in the future.

htcfreek commented on src/modules/cmdpal/ext/Microsoft.CmdPal.Ext.Bookmark/Helpers/OpenInShellHelper.cs:
```diff
@@ -0,0 +1,51 @@
+            process.Start();
+            return true;
+        }
+        catch (Win32Exception ex)
```
Handle userCanceled exception specific as "do nothing". It happens if user cancels the uac dialog.

moooyo replied:
Good suggestion.

htcfreek replied:
where is the change?

htcfreek commented on src/modules/cmdpal/ext/Microsoft.CmdPal.Ext.Bookmark/Helpers/OpenInShellHelper.cs:
Can't we use it from the extension api? Would make sense to move it.

moooyo replied:
yeah, but if we need to change the SDK, we need very carefully to do it because we have the responsibility to maintain in the future.

Currently, I don't think it's necessary to move it in SDK. but I agree to move it in a common position. Seems we don't have it right now.

htcfreek reviewed (changes requested):
I took a look at the code. See my comments.

_I do not have the time to run and try it. And I do not have the permission to unblock your PR later._

Commit:
fix

Commit:
Add error msg as return

zadjii-msft commented:
I don't think I love the type dropdown here. I don't really want people to have to think about what the type of thing is. Plus, if we add built-in support for python 2 and 3, then why not other things? why not bash or node or ...? What if they don't have python installed? Or `pwsh`?

I think it's cleaner for us to require the actual executable in the bookmark. And then we can heuristically figure out:
* Is this a URL -> launch it with the URL opener (we probably could move the Open in browser stuff from the web search plugin to the toolkit)
* Is it a path to a folder? -> add the open folder commands (we probably could even lift some stuff out of the indexer into the toolkit for that)
* else, it's something to `ShellExecute`. (`test.bat`, or `cmd.exe /k ipconfig`, or `pwsh foo.ps1`, or `py foo.py` or `path\to\some_exe.exe with args`, or heck even `"path\to\My Word Doc.docx"` etc) and we should just start that. 

I know that's not necessarily as clean as just explicitly bookmarking "`foo.py` as a Python bookmark", but I think it's almost better to just have to bookmark "`python foo.py`". 

(plus, I bet we could try to use the ThumbnailHelper to get the thumbnail of the exe for the bookmark)

htcfreek commented:
@zadjii-msft 
Fair point. And this solves my short thought about not being able to select "Windows Terminal".

zadjii-msft commented:
off-topic: it might be a hot take, but I don't think we should have the "open in Windows Terminal"-like options anymore. The Default Terminal setting is available back even on Windows 10 these days. We should just always respect that

(I believe I have similar notes in #38277)

Commit:
Merge main

Commit:
Ok, fine. We can detect the bookmark type by ourself. No need to explicitly make customer decide it.

Commit:
Fix typo

Commit:
Fix typo

Commit:
Fix typo

htcfreek commented on src/modules/cmdpal/ext/Microsoft.CmdPal.Ext.Bookmark/Helpers/BookmarkTypeHelper.cs:
```diff
@@ -0,0 +1,94 @@
+            // Otherwise, we assume it's a powershell or pwsh.
+            // Prefer to use pwsh, but check if pwsh is installed
+            // if not, we use powershell
+            if (EnvironmentsCache.Instance.TryGetExecutableFileFullPath("pwsh.exe", out _))
```
isn't this force powershell 5 scripts to run in pwsh 7 if both are installed? Bad idea.

htcfreek replied:
no. don't do this. Because they are not compatible. Well you can decide based on executable used. or not?

htcfreek commented on src/modules/cmdpal/ext/Microsoft.CmdPal.Ext.Bookmark/Helpers/BookmarkTypeHelper.cs:
```diff
@@ -0,0 +1,94 @@
+    {
+        var splittedBookmarkValue = bookmark.Split(" ");
+
+        if (splittedBookmarkValue.Length > 1)
```
this code part is wired. How do you know that it is a powershell command and not e.g. an AutoIt command at this place? 🤔

moooyo replied:
No..
I don't mean it's a powershell command. I mean it's a shell command. Maybe a command prompt or pwsh. Auto It is also a choice for it.

But, how to decide which shell we need? I just make a simple decision strategy but if we have a more better choice, please let me know.

htcfreek replied:
what i mean is: at this point of code you never validated that we call a ps1 file. You only validate for cmd/bat. And therefore it can be an ahk file or something else.

htcfreek commented on src/modules/cmdpal/ext/Microsoft.CmdPal.Ext.Bookmark/Command/ShellCommand.cs:
```diff
@@ -0,0 +1,80 @@
+
+        if (bookmarkType == BookmarkType.Cmd)
+        {
+            args = $"/C {bookmarkValue}";
```
how to keep open if forced to "/c"?

moooyo replied:
Emmm...
How about add a setting for it?

htcfreek commented on src/modules/cmdpal/ext/Microsoft.CmdPal.Ext.Bookmark/Command/ShellCommand.cs:
```diff
@@ -0,0 +1,80 @@
+        }
+        else
+        {
+            args = $"-Command \"{bookmarkValue}\"";
```
can user for to keep open? What if "-File" param is required?

Commit:
Merge branch 'main' into yuleng/cmdpal/bookmarks

Commit:
merge main and add settings

Commit:
store

Commit:
Merge branch 'main' into yuleng/cmdpal/bookmarks

Commit:
simplify the logic. Just call them directly.

Commit:
Remove setting page

Commit:
Remove unused todo

Commit:
Remove unused resource

moooyo commented:
How about the new one? Just call the command directly. So we don't need to consider which command type it is.

@zadjii-msft
@htcfreek

zadjii-msft reviewed (changes requested):
Okay I have feedback that I don't think fits in one place in particular

* I think we can probably make this a lot simpler, by getting rid of the File and Folder command types too. We can probably just `OpenInShellHelper.OpenInShell` them too.
* Trying to determine if it's a "command" or a "path" based on the presence of a literal space is probably not the right way to do that.
  - Consider:
    - `cmd.exe` (this is a "command" that doesn't have a space)
    - `c:\users\Mike\OneDrive - Microsoft\foo.png` (this is a file that has a space in the path)
  - I think we need to better be able to parse the string into (path, args...). I would think that System.CommandLine should be able to do that? But this is notoriously challenging
* We may want to do the lazy ThumbnailHelper icon loading for files/folders/{the exe of a command}, rather than just the static File/Folder/Command icons
  - But that could be just a follow-up
* I only don't immediately recommend merging the web commands with the rest, because that one has Special Favicon Handling
  - by all accounts, OpenInShellHelper.OpenInShell should Just Work for URLs too
  - heck, I bet we could still probably figure out the favicon handling without too much trouble

Commit:
Merge folder and file

Commit:
Fetch icon

Commit:
Remove unused string

moooyo commented on src/modules/cmdpal/ext/Microsoft.CmdPal.Ext.Bookmark/Properties/Resources.resx:
```diff
@@ -161,4 +161,37 @@
+  </data>
+  <data name="bookmarks_form_bookmark_type_CMD" xml:space="preserve">
+    <value>Command Prompt</value>
+  </data>
```
Yes.

Removed

moooyo commented on src/modules/cmdpal/ext/Microsoft.CmdPal.Ext.Bookmark/Pages/AddBookmarkPage.cs:
```diff
@@ -2,6 +2,7 @@
 // The Microsoft Corporation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using Microsoft.CmdPal.Ext.Bookmarks.Models;
```
No, this one we still need it.

Commit:
fix typo

Commit:
Merge urlCommand into shellCommand

Commit:
Add folder context menu back

moooyo commented:
> Okay I have feedback that I don't think fits in one place in particular
> 
> * I think we can probably make this a lot simpler, by getting rid of the File and Folder command types too. We can probably just `OpenInShellHelper.OpenInShell` them too.
> * Trying to determine if it's a "command" or a "path" based on the presence of a literal space is probably not the right way to do that.
>   
>   * Consider:
>     
>     * `cmd.exe` (this is a "command" that doesn't have a space)
>     * `c:\users\Mike\OneDrive - Microsoft\foo.png` (this is a file that has a space in the path)
>   * I think we need to better be able to parse the string into (path, args...). I would think that System.CommandLine should be able to do that? But this is notoriously challenging
> * We may want to do the lazy ThumbnailHelper icon loading for files/folders/{the exe of a command}, rather than just the static File/Folder/Command icons
>   
>   * But that could be just a follow-up
> * I only don't immediately recommend merging the web commands with the rest, because that one has Special Favicon Handling
>   
>   * by all accounts, OpenInShellHelper.OpenInShell should Just Work for URLs too
>   * heck, I bet we could still probably figure out the favicon handling without too much trouble

Emm, you are right. 


Made the following changes:
1. Merged the file, folder and Web into OpenInShell. But keep the bookmark type in to avoid dup jug. Yeah, you are right. We can merge all of them into openInShell. Actually, favicon is not a blocker. I've moved this logic into IconHelper.
2. Fetch file and folder icon (ha actually I want to create a new PR for it before but we need a big change for this PR. So I decide to do it in this PR)
3. Remove unused string.
4. Updated the PR's descrption.

moooyo commented:
> 1. Merged the file, folder

The only bug is we can not open a "oneDrive" folder as I did in the PR's descrption. But the normal folder is ok.

Commit:
update comments

Commit:
Merge branch 'main' into yuleng/cmdpal/bookmarks

moooyo commented:
/azp run

yeelam-gordon commented on src/modules/cmdpal/ext/Microsoft.CmdPal.Ext.Bookmark/Helpers/BookmarkTypeHelper.cs:
```diff
@@ -0,0 +1,46 @@
+    public static BookmarkType GetBookmarkTypeFromValue(string bookmark)
+    {
+        // judge if the bookmark is a url
+        if (Uri.TryCreate(bookmark, UriKind.Absolute, out var uriResult))
```
In the past, if I type "www.bing.com", it will open and save as URL in browser for me.
Probably because the "Default type" is Web rather than Command.

Will it be a problem now?

moooyo replied:
> In the past, if I type "www.bing.com", it will open and save as URL in browser for me. Probably because the "Default type" is Web rather than Command.
> 
> Will it be a problem now?

yeah, you catch it.

yeelam-gordon commented on src/modules/cmdpal/ext/Microsoft.CmdPal.Ext.Bookmark/Command/ShellCommand.cs:
```diff
@@ -0,0 +1,88 @@
+            ExtensionHost.LogMessage($"Failed to open {bookmarkValue} in shell. Ex: {errorMessage}");
+            return CommandResult.ShowToast(new ToastArgs() { Message = Resources.bookmarks_command_invoke_failed_message });
+        }
+
```
These code... has too many assumptions...
I remember someone suggest to use CommandLine parser before which is avaliable in .net? ( I remember I did use it for my cli program in the past).

yeelam-gordon commented on src/modules/cmdpal/ext/Microsoft.CmdPal.Ext.Bookmark/Pages/AddBookmarkForm.cs:
```diff
@@ -73,33 +75,14 @@ public override CommandResult SubmitForm(string payload)
         updated.Bookmark = formBookmark.ToString();
-        updated.Type = bookmarkType;
+        updated.Type = parsedBookmarkType;
+        var paths = EnvironmentsCache.Instance.GetPaths();
```
You got the paths, and not using it?
I think it will combine for open command or open file later on?

moooyo replied:
Removed. Unused code.

yeelam-gordon commented:
I think it is difficult to "assume" what's URL, File/Folder, or Command.
If we add more thing later on, it just not scalability here.

Can we have "interaction" between extension? i.e. At the "Run Command", I can add say "Bookmark" it?
Since people can try out the command at the "Run command", and then bookmark it.

yeelam-gordon reviewed (commented):
Not suggest to get in for 0.91 except there is special reason

yeelam-gordon commented:
> I think it is difficult to "assume" what's URL, File/Folder, or Command. If we add more thing later on, it just not scalability here on what does a Command parsing means.
> 
> Two thoughts:
> 
> 1. Can we have "interaction" between extension? i.e. At the "Run Command", I can add say "Bookmark" it?
>    Since people can try out the command at the "Run command", and then bookmark it.
> 2. It will be more streamline if we guess the type first, and then let people to "change" the bookmark type later on.

Looks like from what Mike just added the issue thread #39261 , bookmark from other common extension like "All Apps", "Search Files (and Folder)", "Run Command", does make sense.
There is already alias and shortcut management for bookmark, just not easy to be discovered at the bookmark UX itself (i.e. need go to setting page now)

moooyo commented:
> > I think it is difficult to "assume" what's URL, File/Folder, or Command. If we add more thing later on, it just not scalability here on what does a Command parsing means.
> > Two thoughts:
> > 
> > 1. Can we have "interaction" between extension? i.e. At the "Run Command", I can add say "Bookmark" it?
> >    Since people can try out the command at the "Run command", and then bookmark it.
> > 2. It will be more streamline if we guess the type first, and then let people to "change" the bookmark type later on.
> 
> Looks like from what Mike just added the issue thread #39261 , bookmark from other common extension like "All Apps", "Search Files (and Folder)", "Run Command", does make sense. There is already alias and shortcut management for bookmark, just not easy to be discovered at the bookmark UX itself (i.e. need go to setting page now)

Yeah, agree.

Commit:
Remove unused code

Commit:
Remove unused code

Commit:
Fix bug

moooyo commented:
> Not suggest to get in for 0.91 except there is special reason

If possible, I want to get it in.  But if not, it's ok to hold.

In fact, the original version also have many assumption. This PR at least I think it just add a new ability to invoke command.

So, If we want to consider how to get the correct bookmark type (through a new choices input or other way), we can do in the future.

Commit:
merge main

Commit:
Remove unused code

Commit:
Remove unused code

Commit:
Improve robust

moooyo commented:
Close it because we have a better one. #40430
